### PR TITLE
Motif2factors from orthologs

### PR DESCRIPTION
**Pull request by Maarten-vd-Sande**
Get a motif2factors file based on orthologs. My idea would be to build this into gimmemotifs, as a new command.  

To do:

- [x] name of command, e.g. `gimme motif2factors`
- [x] which species as (default) reference (orthodb advices somewhere between 4-8 I believe)
- [x] test if this also works on non-gimmemotifs database (e.g. jaspar)
- [x] test on ananse + docs for ananse (https://github.com/vanheeringen-lab/ANANSE/projects/4#card-54237974)
- [x] clean up code
  - [x] generic
  - [x] how do we log for shell commands, should I use the logger?
- [x] docs 
- [ ] update the current gimme motif database and remove the _unclear_ factors 
- [ ] support non-vertebrate?

### CODE REVIEW

**Comment by Maarten-vd-Sande** [comment]
Of these factors I can not find a protein in the gtf file (or are not in the "blacklist"):

```
ENSG00000234254
ENSMUSG00000044690
ENSG00000235187
AC012531.1
EBOX
Gm4830
DPRX
Q8VHG7_MOUSE
Gm5889
ENSMUSG00000040953
ENSG00000187728
ESX1
AL662830.5
AL845464.3
BX927239.6
CR933878.8
BX284686.1
CR812478.2
ENSG00000250096
AL773544.1
CR759815.4
CR388229.2
BX088580.2
AL662833.4
Rhox8
IRC900814
CR354565.3
CR759733.4
AL844527.7
AL662824.5
CR936877.3
ENSMUSG00000090020
CACBP
AC002126.6
ENSMUSG00000079994
mix-a
ENSG00000250811
Rbpsuh-rs3
AC189028.1
NPAS
CACD
Gm5454
TWIST1
HINFP1
Twist1
AL844853.8
CR388219.4
CR759784.3
AL662834.13
ENSG00000233294
ENSG00000215754
ENSG00000236207
ENSG00000228710
ENSG00000232923
DUX4L3
DUX4L5
DUX4L6
ENSG00000233794
ENSG00000225958
ENSG00000230380
ENSG00000233361
ENSG00000224558
ENSG00000237516
ENSG00000234550
ENSG00000250589
DUX4L2
ENSG00000226866
ENSG00000227522
CT476828.2
ENSG00000217642
AL592170.2
ENSG00000233698
ENSG00000230776
NR1H
EWSR1-FLI1
Srebp1a
CR925796.2
CR753803.1
AL662828.6
AC226150.2
Zfp57
SEF1
GTF2A
COMP1
HF1H3B
ENSG00000250542
Esx1
```

The dux4lN factors are pseudogenes, and are not taken along by gffread because of this. I don't know if that is a problem or not

**Comment by Maarten-vd-Sande** [comment]
@simonvh could you take a generic look if this seems okay? The current code is how I used it last, and this seems to produce okay results.

Do you have any thoughts on the checkboxes above?

File: gimmemotifs/orthologs.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+import pandas as pd
+
+# TODO: actually use logger
+logger = logging.getLogger("gimme.orthologs")
+
+FASTA_LINEWIDTH = 80
+BLACKLIST_TFS = [
+    "Dobox4",  # does not exist
+    "Dobox5",  # does not exist
+    "FOXA",  # not sure which FOXA(1,2,3)
```
Comment (simonvh) [inline gimmemotifs/orthologs.py:34]: EP300

It's not in the gimme database, but may be present in other databases.

File: gimmemotifs/orthologs.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+import logging
+import genomepy
+import numpy as np
+import pandas as pd
+
+# TODO: actually use logger
+logger = logging.getLogger("gimme.orthologs")
+
+FASTA_LINEWIDTH = 80
+BLACKLIST_TFS = [
```
Comment (simonvh) [inline gimmemotifs/orthologs.py:31]: Make blacklist tfs case insensitive?

File: gimmemotifs/orthologs.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+    outfile, new_reference, database_references, motifsandfactors, database
+):
+    """
+    Make a motifs2factors file based on an existing ortolog database.
+
+    We loop over each motif, find which factors belong to it in the old
+    reference, then find all the orthogroups those belong to, and finally assign
+    all the genes of the new reference species that belong to those orthogroups.
+
+    However
```
Comment (simonvh) [inline gimmemotifs/orthologs.py:345]: Oooh.. exciting! Continued in part II?

File: gimmemotifs/orthologs.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    """
+    Returns all the orthogroups a factor belongs to. This is NOT a trivial task.
+
+    Some of the factors shouldn't be considered (blacklisted), some have an
+    ambiguous name that should be renamed, and in general the factor naming is
+    inconsistent.
+
+    We first check whether or not it fits in our ortogroup database (easy!), if
+    not, we start querying mygeneinfo to get more gene aliases to search for in
+    our ortogroups. If that didn't work we query mygeneinfo again, but with a
```
Comment (simonvh) [inline gimmemotifs/orthologs.py:396]: Maybe make this configurable? Have 3 documented settings, something like `exact`, `strict` and `lenient`. Although `strict` may not be the best term.

**Review by simonvh — APPROVED** [review]

**Comment by simonvh** [comment]
> * [ ]  name of command, e.g. `gimme motif2factors`

Ideally I would like something shorter, but this is descriptive. Ok with me.

> * [ ]  which species as (default) reference (orthodb advices somewhere between 4-8 I believe)

Selection of species roughly make sense. Zebrafish and medaka are enough as standard fished, for some species we should check if there's not a better quality genome in the general phylogenetic neighborhood.

> * [ ]  test if this also works on non-gimmemotifs database (e.g. jaspar)

Yes.

>   * [ ]  how do we log for shell commands, should I use the logger?
Catch output, log using loguru.

> * [ ]  update the current gimme motif database and remove the _unclear_ factors

Yes, this is a little more work though. Maybe not for now.

> * [ ]  support non-vertebrate?

I think it should work, by choosing the right reference genomes? Just a case of documentation. I don't know enough about plant or other invertebrate motifs to check if it is reasonable.

File: gimmemotifs/orthologs.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+    Finally, based on these ortologs, a new motifs2factors file is created from
+    the old one.
+    """
+    all_genomes = set(database_references + extra_orthologs_references + new_reference)
+
+    # TODO use tmpdir
+    # TODO set outdir from args
+    # outdir = tempfile.mkdtemp(dir=tmpdir)
+    outdir = "/vol/gimmetest"
+    genomes_dir = genomes_dir if genomes_dir is not None else outdir
```
[the remark this answers is not in the document]
Comment (Maarten-vd-Sande) [reply gimmemotifs/orthologs.py:78]: I am not sure if the user actually wants these genomes (apart from the one based on the new reference)

[the remark this answers is not in the document]
Comment (Maarten-vd-Sande) [reply gimmemotifs/orthologs.py:78]: For example on cn45/106 there is only 5 gb of space on home. If we were to download 8 genomes there it is full.. My idea was to put this in tmp (not outdir as the code suggests)

[the remark this answers is not in the document]
Comment (simonvh) [reply gimmemotifs/orthologs.py:78]: Yeah, you're correct. But you may already have some of them, and then it's extra space. Would it be possible combine it? Use genomes from the genomepy dir if present, and add genomes to the tmp dir that are not.

[the remark this answers is not in the document]
Comment (Maarten-vd-Sande) [reply gimmemotifs/orthologs.py:78]: Anything is possible :smile:

**Comment by Maarten-vd-Sande** [comment]
Seems to work decent for JASPAR (at least if I check if names overlap):
http://ocimum.science.ru.nl/GRCz11.JASPAR2020_vertebrates.motif2factors.txt

It has problems at the bottom with all the specific Zinc Fingers. The problem with these is that they all get put in the same orthogroup, and have no relation to other species, e.g.:

```
.|ENSG00000283201, ZNF724|ENSG00000196081, ZNF708|ENSG00000182141, ZNF92|ENSG00000146757, ZNF98|ENSG00000197360, ZNF492|ENSG00000229676, ZNF676|ENSG00000196109, ZNF208|ENSG00000160321, ZNF728|ENSG00000269067, ZNF726|ENSG00000213967, ZNF91|ENSG00000167232, ZNF254|ENSG00000213096, ZNF680|ENSG00000173041, ZNF729|ENSG00000196350, ZNF99|ENSG00000213973, ZNF493|ENSG00000196268, ZNF430|ENSG00000118620, ZNF738|ENSG00000172687, ZNF431|ENSG00000196705, ZNF100|ENSG00000197020, ZNF730|ENSG00000183850, ZNF273|ENSG00000198039, ZNF735|ENSG00000223614, ZNF722P|ENSG00000241149, ZNF479|ENSG00000185177, ZNF716|ENSG00000182111, ZNF679|ENSG00000197123, ZNF595|ENSG00000272602, ZNF141|ENSG00000131127, ZNF732|ENSG00000186777, ZNF718|ENSG00000250312, ZNF736|ENSG00000234444, ZNF727|ENSG00000214652, ZNF695|ENSG00000197472, ZNF682|ENSG00000197124, ZNF195|ENSG00000005801, ZNF721|ENSG00000182903, ZNF678|ENSG00000181450, ZNF117|ENSG00000152926, ZNF714|ENSG00000160352, ZNF85|ENSG00000105750, ZNF723|ENSG00000268696, ZNF138|ENSG00000197008, ZNF107|ENSG00000196247, ZNF681|ENSG00000196172, ZNF43|ENSG00000198521, ZNF675|ENSG00000197372, ZNF429|ENSG00000197013, ZNF90|ENSG00000213988, ZNF506|ENSG00000081665, ZNF486|ENSG00000256229, ZNF66|ENSG00000160229, ZNF93|ENSG00000184635, ZNF737|ENSG00000237440, ZNF626|ENSG00000188171, ZNF257|ENSG00000197134, ZNF253|ENSG00000256771
```

Which has no hits with other species..

**Comment by simonvh** [comment]
> Seems to work decent for JASPAR (at least if I check if names overlap):
> http://ocimum.science.ru.nl/GRCz11.JASPAR2020_vertebrates.motif2factors.txt
> 
> It has problems at the bottom with all the specific Zinc Fingers. The problem with these is that they all get put in the same orthogroup, and have no relation to other species, e.g.:
> 
> ```
> .|ENSG00000283201, ZNF724|ENSG00000196081, ZNF708|ENSG00000182141, ZNF92|ENSG00000146757, ZNF98|ENSG00000197360, ZNF492|ENSG00000229676, ZNF676|ENSG00000196109, ZNF208|ENSG00000160321, ZNF728|ENSG00000269067, ZNF726|ENSG00000213967, ZNF91|ENSG00000167232, ZNF254|ENSG00000213096, ZNF680|ENSG00000173041, ZNF729|ENSG00000196350, ZNF99|ENSG00000213973, ZNF493|ENSG00000196268, ZNF430|ENSG00000118620, ZNF738|ENSG00000172687, ZNF431|ENSG00000196705, ZNF100|ENSG00000197020, ZNF730|ENSG00000183850, ZNF273|ENSG00000198039, ZNF735|ENSG00000223614, ZNF722P|ENSG00000241149, ZNF479|ENSG00000185177, ZNF716|ENSG00000182111, ZNF679|ENSG00000197123, ZNF595|ENSG00000272602, ZNF141|ENSG00000131127, ZNF732|ENSG00000186777, ZNF718|ENSG00000250312, ZNF736|ENSG00000234444, ZNF727|ENSG00000214652, ZNF695|ENSG00000197472, ZNF682|ENSG00000197124, ZNF195|ENSG00000005801, ZNF721|ENSG00000182903, ZNF678|ENSG00000181450, ZNF117|ENSG00000152926, ZNF714|ENSG00000160352, ZNF85|ENSG00000105750, ZNF723|ENSG00000268696, ZNF138|ENSG00000197008, ZNF107|ENSG00000196247, ZNF681|ENSG00000196172, ZNF43|ENSG00000198521, ZNF675|ENSG00000197372, ZNF429|ENSG00000197013, ZNF90|ENSG00000213988, ZNF506|ENSG00000081665, ZNF486|ENSG00000256229, ZNF66|ENSG00000160229, ZNF93|ENSG00000184635, ZNF737|ENSG00000237440, ZNF626|ENSG00000188171, ZNF257|ENSG00000197134, ZNF253|ENSG00000256771
> ```
> 
> Which has no hits with other species..

Zinc fingers  proteins are very species-specific. No surprise that they're not found in other species (and therefore not a significant issue).

**Review by simonvh — APPROVED** [review]

**Comment by Maarten-vd-Sande** [comment]
@simonvh this is ready to be merged. Should be ready for use

**Pull request merged**